src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,132 @@
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> currentNode = tail;
+        Node<T> newNode = new Node<>(currentNode, value, null);
+        tail = newNode;
+        if (head == null) {
+            head = newNode;
+        } else {
+            currentNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> currentNode = findByIndex(index);
+        Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+        if (index == 0) {
+            head = newNode;
+        } else {
+            currentNode.prev.next = newNode;
+        }
+        currentNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = findByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = findByIndex(index);
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == object || currentNode.value != null
+                    && currentNode.value.equals(object)) {
+                remove(i);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findByIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index + " Size: " + size);
+        }
+        int currentIndex = 0;
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (currentIndex == index) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+            currentIndex++;
+        }
+        return head;
+    }
+
+    private void unlink(Node<T> currentNode) {
+        if (currentNode == head) {
+            head = currentNode.next;
+            if (size > 1) {
+                head.prev = null;
+            }
+        } else if (currentNode == tail) {
```

Comment:
Let's try to refactor the whole `if else` block in this method not by checking `size` and `== head`, but checking `currentNode.prev == null` `currentNode.next == null` etc.
You also can introduce a new variables for better readability e.g.:
```
   final T element = currentNode.value;
        final Node<T> next = currentNode.next;
        final Node<T> prev = currentNode.prev;
```

Reply:
he probably meant to record something like this. it would also be good to make variables for currentNode.next and currentNode.prev. Maybe I got the names wrong
```suggestion
       if (currentNode.prev == null) {
            first = currentNode.next;
        } else {
            currentNode.prev.next = currentNode.next;
        }
        if (currentNode.next == null) {
            last = currentNode.prev;
        } else {
            currentNode.next.prev = currentNode.prev;
        }
```